### PR TITLE
PERF-1976 Add error and retry counters for MongoDB

### PR DESCRIPTION
**Pull request by mdcallag**

### CODE REVIEW

File: src/main/java/com/facebook/LinkBench/LinkStoreMongoDb2.java [hunk 1]
```diff
@@ -1055,27 +1079,35 @@ public String getName()  {
...
+                        if (retries >= max_retries || !isRetryableError(e.getCode())) {
+                            logger.error(task.getName() + " failed after " + max_retries + " retries.", e);
                             throw e;
                         }
-                        logger.debug(task.getName() + "("+ retries + ") retrying " + e.getMessage());
+                        logger.warn(task.getName() + "("+ retries + ") retrying " + e.getMessage());
+                        counters[0].incrementAndGet();
+                        // This has a race between get and set. It isn't worth preventing that race.
+                        if (retries > counters[1].get())
+                            counters[1].set(retries);
```
Comment (henrikingo) [inline src/main/java/com/facebook/LinkBench/LinkStoreMongoDb2.java]: Could the counters be something else than 0 and 1? I believe in Java you can use Enum or just a final variable to name the indexes something.

File: src/main/java/com/facebook/LinkBench/LinkStoreMongoDb2.java [hunk 2]
```diff
@@ -314,13 +315,28 @@ private boolean isRetryableError(int errCode) {
 
     @Override
     public LinkWriteResult addLink(final String dbid, final Link link, final boolean noinverse) {
-        logger.debug("addLink " + link.id1 +
-            "." + link.id2 +
-            "." + link.link_type);
+        logger.debug("addLink " + link.id1 + "." + link.id2 + "." + link.link_type);
+        return newLink(dbid, link, noinverse, retry_add_link, max_add_link, "addLink", true);
+    }
```
[the remark this answers is not in the document]
Comment (mdcallag) [reply src/main/java/com/facebook/LinkBench/LinkStoreMongoDb2.java]: I don't think so, but not sure I understand. Can you be more specific?

File: src/main/java/com/facebook/LinkBench/LinkStoreMongoDb2.java (hunk 1 above)
Comment (mdcallag) [reply src/main/java/com/facebook/LinkBench/LinkStoreMongoDb2.java]: That would work if I were using a map and then the enum values could be keys for the map. But this is an array. Java doesn't allow inner classes to reference non-final fields from outer classes so it was difficult to figure out how to pass the counters into this code.

File: src/main/java/com/facebook/LinkBench/LinkStoreMongoDb2.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (henrikingo) [reply src/main/java/com/facebook/LinkBench/LinkStoreMongoDb2.java]: Ah, the opening bracket just happened to wrap on my screen. Never mind.

File: src/main/java/com/facebook/LinkBench/LinkStoreMongoDb2.java (hunk 1 above)
Comment (henrikingo) [reply src/main/java/com/facebook/LinkBench/LinkStoreMongoDb2.java]: Couldn't you just do 

    final int ERRORS = 0;
    final int RETRIES = 1;

locally in this function?

Comment (mdcallag) [reply src/main/java/com/facebook/LinkBench/LinkStoreMongoDb2.java]: done

**Pull request merged**